### DISCUSSION TITLE
How to use `useMeta` only in backend

### DISCUSSION THREAD

**Discussion by djschilling** [nuxt/nuxt#17928] [Questions] [answered]
I have this code in one of my vue pages:

```
<script lang="ts" setup>
import {useNuxtApp} from "#app";

const nuxtApp = useNuxtApp();

const url = nuxtApp?.ssrContext?.url;
useMeta({
    meta: [
        { name: 'og:url', content: url }
    ]
});
</script>
```

The code gets correctly executed in the backend and sets the `meta` element.
But then the frontend renders in the browser, and there `ssrContext` ist not set and so the url is set back to `undefined`.

What i want is to only set the Meta Tags in The backend. Is there a way to do this?

**Comment by viii47** [comment]
> What i want is to only set the Meta Tags in The backend. Is there a way to do this?

There's no need to do this. If you have dynamically changing content you can `compute()` the value.

### Example
```js
const url = nuxtApp?.ssrContext?.url;
useMeta({
    meta: [
        { name: 'og:url', content: computed(() => url) }
    ]
});
```

**Comment by Diizzayy** [comment, marked as answer]
> But then the frontend renders in the browser, and there ssrContext ist not set and so the url is set back to undefined

@djschilling Ideally you'd be passing your data to the page via `useAsyncData`, you can also use it to achieve what you want as seen below

```js
<script lang="ts" setup>
const { data } = useAsyncData("ogurl", async () => {
  return { url: useNuxtApp().ssrContext.url };
});

useMeta({
  meta: [
    {
      hid: "og:url",
      name: "og:url",
      content: data.value.url,
    },
  ],
});
</script>
```

Also, the og:url should be an absolute url to your page, as opposed to the relative url given from `ssrContext.url`.

The code below is a snippet from my project that should be perfect for your SEO needs

create a new directory in your project called `composables`, within that folder create a new file called `seo.ts` then add the paste the code below in the file.

`composables/seo.ts`
```js
import { MetaObject } from 'nuxt3/dist/meta/runtime'

type SEOData = {
  title: string
  desc?: string
  image?: string
  type?: string
}

export const useSEO = () => {
  let domain = useState('domain', () => 'https://example.com')
  const route = useRoute()

  const image = 'https://example.com/image.jpg'

  if (process.server) {
    const { ssrContext } = useNuxtApp()

    const host = ssrContext.req?.headers?.host
    const protocol = ssrContext.req?.headers?.referer?.split('/')?.[0]

    if (host) {
      if (!protocol) domain.value = `https://${host}`
      else domain.value = `${protocol}//${host}`
    }
  }

  function GenerateTags(data: SEOData): MetaObject {
    const title = `${data.title} | Diizzayy`
    const canonical = `${domain.value}${route.fullPath}`

    const link = [
      { rel: 'canonical', href: canonical },
      { rel: 'alternate', href: canonical, hreflang: 'en' },
    ]

    const opengraph = [
      { hid: 'og:url', property: 'og:url', content: canonical },
      { hid: 'og:locale', property: 'og:locale', content: 'en_US' },
      { hid: 'og:type', property: 'og:type', content: data.type || 'website' },
      { hid: 'og:title', property: 'og:title', content: title },
      { hid: 'og:description', property: 'og:description', content: data.desc },
      {
        hid: 'og:site_name',
        property: 'og:site_name',
        content: 'Diizzayy',
      },
      { hid: 'og:image:alt', property: 'og:image:alt', content: title },
      {
        hid: 'og:image',
        property: 'og:image',
        content: data.image || image,
      },
    ]

    const twitter = [
      {
        hid: 'twitter:card',
        property: 'twitter:card',
        content: 'summary_large_image',
      },
      { hid: 'twitter:title', property: 'twitter:title', content: title },
      {
        hid: 'twitter:description',
        property: 'twitter:description',
        content: data.desc,
      },
      {
        hid: 'twitter:image',
        property: 'twitter:image',
        content: data.image || image,
      },
      {
        hid: 'twitter:image:alt',
        property: 'twitter:image:alt',
        content: title,
      },
    ]

    useMeta ({
      title,
      link,
      meta: [
        {
          hid: 'description',
          property: 'description',
          content: data.desc,
        },
        ...opengraph,
        ...twitter,
      ],
    })
  }

  return { GenerateTags }
}
```

you can now use this composable in your pages as seen below

```js
<script lang="ts" setup>
const { GenerateTags } = useSEO()

GenerateTags({ title: 'Home' })
</script>
```

**Reply by djschilling** [reply]
Thanks for the information. This is really helpful.

I am aware of the url problem but just wanted to provide a simple example.

**Reply by Diizzayy** [reply]
No problem, I'm happy to help. Be sure to make this question as answered if this resolved it for you.

**Reply by sewalsh** [reply]
This is great @Diizzayy! However, I notice the meta doesn't update when navigating client-side (after the initial SSR request). I've tried using computed/watch but no luck. Were you able to get it working?

**Reply by Diizzayy** [reply]
@sewalsh simply calling `useSEO().GenerateTags({})` is usually suffice, however though there appears to be a bug in the current version of Nuxt3 that prevents subsequent meta updates ( after initial SSR rendering ) from being applied upon client-side navigation.

**Reply by Diizzayy** [reply]
@sewalsh This is now fixed as of #3271. Thanks for pointing out the issue!

**Reply by sewalsh** [reply]
Cheers! I can confirm it's working as expected now. :)
